check/controls.go

```diff
@@ -79,12 +74,85 @@ func NewControls(in []byte, definitions []string) (*Controls, error) {
 	return c, nil
 }
 
+func (controls *Controls) convertAuditToRegisteredType(auditType AuditType, audit interface{}) (Auditer, error) {
+
+	if auditType == "" || auditType == TypeAudit {
+
+		if s, ok := audit.(string); ok || audit == nil {
+			return &Audit{audit: s}, nil
+		}
+		return nil, fmt.Errorf("failed to convert audit, mismatching type") // TODO add Check ID here
+	} else {
+
+		if callback, ok := controls.auditTypeRegistry[auditType]; ok {
+
+			o := callback()
+			if auditAsString, err := yaml.Marshal(audit); err == nil {
+
+				if err := yaml.Unmarshal(auditAsString, o); err != nil {
+					return nil, fmt.Errorf("unable to Unmarshal Audit %v", err)
+
+				}
+				return o.(Auditer), nil // we don't check error because it already verified in convertAuditToRegisteredType
+
+			} else { // 'else' clause used here due to 'err' variable scope
+				return nil, fmt.Errorf("unable to marshal Audit %v", err)
+			}
+		}
+		return nil, fmt.Errorf("audit type %v is not registered", auditType)
+	}
+}
+
+func (controls *Controls) RegisterAuditType(auditType AuditType, typeCallback func() interface{}) error {
+
+	if _, ok := controls.auditTypeRegistry[auditType]; !ok {
+		a := typeCallback()
+		if _, ok := a.(Auditer); !ok {
+			return fmt.Errorf("audit type %v must implement Auditer interface", auditType)
+		}
+		controls.auditTypeRegistry[auditType] = typeCallback
+		return nil
+
+	}
+	return fmt.Errorf("audit type %v already registered", auditType)
+}
+
+func extractAllAudits(controls *Controls) (err error) {
+	// Prepare audit commands
+	for _, group := range controls.Groups {
+		for _, check := range group.Checks {
+			if check.SubChecks == nil {
+				if a, err := controls.convertAuditToRegisteredType(check.AuditType, check.Audit); err == nil {
+
+					check.auditer = a
```

Comment:
@igorbeider I don't know what has happened to the comments I made on this yesterday, but I still don't think this will work. `check` is *not* a pointer to the element in `group.Checks`, it is a *copy*. See https://tour.golang.org/moretypes/16

Reply:
Ah, OK now I see - you have made it a slice of pointers to make this work! :-D 

This is not a very Go-like way to deal with the problem. A much more idiomatic way would be to use the index from the range, and modify the actual item in the array.